### PR TITLE
DIV-6561-Adding-new-Service-Application-documents-collection-definition

### PR DESCRIPTION
**Pull request by madjava**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/DIV-6561 a subtask of https://tools.hmcts.net/jira/browse/DIV-6532

### Change description ###

Adding new CCD collection to hold multiple service application documents. 
Enhancements to CaseTypeTab unit test suite

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json [hunk 1]
```diff
@@ -705,5 +705,47 @@
...
+    "CaseFieldID": "ServiceApplicationDocuments",
+    "UserRole": "caseworker-divorce-solicitor",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "03/09/2020",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "ServiceApplicationDocuments",
+    "UserRole": "citizen",
+    "CRUD": "CRU"
```
Comment (madjava) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json:749]: Adding permissions to user roles

File: definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json [hunk 2]
```diff
@@ -1,4 +1,13 @@
 [
+  {
+    "LiveFrom": "24/09/2020",
+    "CaseTypeID": "DIVORCE",
+    "ID": "ServiceApplicationDocuments",
+    "Label": "Service Application Documents",
+    "FieldType": "Collection",
+    "FieldTypeParameter": "DivorceServiceApplicationDocument",
+    "SecurityClassification": "Public"
+  },
```
Comment (madjava) [inline definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json:10]: New document collection

File: definitions/divorce/json/CaseTypeTab/CaseTypeTab-deemed-and-dispensed-nonprod.json [hunk 3]
```diff
@@ -1,4 +1,44 @@
...
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "DIVORCE",
+    "Channel": "CaseWorker",
+    "TabID": "documents",
+    "TabLabel": "Documents",
+    "TabDisplayOrder": 10,
+    "CaseFieldID": "DocumentsUploadedDnClarification",
+    "TabFieldDisplayOrder": 6
+  },
```
Comment (madjava) [inline definitions/divorce/json/CaseTypeTab/CaseTypeTab-deemed-and-dispensed-nonprod.json:41]: Based on AC, adding new collection display location index

File: definitions/divorce/json/CaseTypeTab/CaseTypeTab-prod.json [hunk 4]
```diff
@@ -1 +1,32 @@
...
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "DIVORCE",
+    "Channel": "CaseWorker",
+    "TabID": "documents",
+    "TabLabel": "Documents",
+    "TabDisplayOrder": 10,
+    "CaseFieldID": "DocumentsUploadedDnClarification",
+    "TabFieldDisplayOrder": 5
+  }
+]
```
Comment (madjava) [inline definitions/divorce/json/CaseTypeTab/CaseTypeTab-prod.json:32]: Moving these here as updated location in non-prod feature file

File: definitions/divorce/json/CaseTypeTab/CaseTypeTab.json [hunk 5]
```diff
@@ -2235,36 +2235,6 @@
...
-  {
-    "LiveFrom": "01/01/2017",
-    "CaseTypeID": "DIVORCE",
-    "Channel": "CaseWorker",
-    "TabID": "documents",
-    "TabLabel": "Documents",
-    "TabDisplayOrder": 10,
-    "CaseFieldID": "DocumentsUploadedDnClarification",
-    "TabFieldDisplayOrder": 5
-  },
```
Comment (madjava) [inline definitions/divorce/json/CaseTypeTab/CaseTypeTab.json:2267]: Ditto

File: definitions/divorce/json/ComplexTypes/ComplexTypes-deemed-and-dispensed-nonprod.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+[
+  {
+    "LiveFrom": "24/09/2020",
+    "ID": "DivorceServiceApplicationDocument",
+    "ListElementCode": "Document",
+    "FieldType": "DivorceDocument",
+    "ElementLabel": "Service Application",
+    "SecurityClassification": "Public"
+  }
+]
```
Comment (madjava) [inline definitions/divorce/json/ComplexTypes/ComplexTypes-deemed-and-dispensed-nonprod.json:10]: Defining collection tpe

File: test/unit/definitions/divorce/CaseTypeTab.test.js [hunk 7]
```diff
@@ -33,117 +48,145 @@ describe('CaseTypeTab', () => {
...
-      const allTabFieldDisplayOrderNumbers = map(allFieldsPerTab, field => {
-        return field.TabFieldDisplayOrder;
-      }).sort((a, b) => {
-        return a - b;
-      });
-      for (let i = 1; i < allTabFieldDisplayOrderNumbers.length; i++) {
-        if (allTabFieldDisplayOrderNumbers[i] - allTabFieldDisplayOrderNumbers[i - 1] !== 1) {
-          assert.fail(`Missing/unordered TabFieldDisplayOrder sequence number in TabID ${tabId} - expected ${allTabFieldDisplayOrderNumbers[i - 1] + 1} but got ${allTabFieldDisplayOrderNumbers[i]}`);
-        }
-      }
```
Comment (madjava) [inline test/unit/definitions/divorce/CaseTypeTab.test.js:65]: Refactored to common location

File: test/unit/definitions/divorce/CaseTypeTab.test.js [hunk 8]
```diff
@@ -33,117 +48,145 @@ describe('CaseTypeTab', () => {
...
-    confidentialCoRespondentCourtAdminLa: 19,
-    notes: 20,
-    marriageCertificate: 21,
-    coRespondent: 22,
-    serviceApplication: 23,
-    LinkedCase: 24,
-    Language: 25,
-    ConfidentialDocuments: 26,
-    General: 1
-  };
```
Comment (madjava) [inline test/unit/definitions/divorce/CaseTypeTab.test.js:97]: Moved to common location

File: test/unit/definitions/divorce/CaseTypeTab.test.js [hunk 9]
```diff
@@ -33,117 +48,145 @@ describe('CaseTypeTab', () => {
...
-      'confidentialCoRespondentCourtAdminLa',
-      'notes',
-      'marriageCertificate',
-      'coRespondent',
-      // 'SolicitorCoRespondent', - TODO - uncomment this when we go live with AOS pack 2
-      'serviceApplication',
-      'LinkedCase',
-      'Language',
-      'ConfidentialDocuments'
-    ]);
```
Comment (madjava) [inline test/unit/definitions/divorce/CaseTypeTab.test.js:139]: Moved to common location

File: test/unit/utils/caseTabTypeHelper.js [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+module.exports = {
+  nonProdTabDisplayOrder,
+  nonProdTabIds,
+  prodTabDisplayOrder,
+  prodTabIds,
+  validateUniqueTabDisplayOrder,
+  validateTabFieldDisplayOrder
+};
```
Comment (madjava) [inline test/unit/utils/caseTabTypeHelper.js:189]: Helper class for case tab type test suite

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json [hunk 11]
```diff
@@ -705,5 +705,47 @@
...
+    "CaseFieldID": "ServiceApplicationDocuments",
+    "UserRole": "caseworker-divorce-courtadmin",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "03/09/2020",
+    "CaseTypeID": "DIVORCE",
+    "CaseFieldID": "ServiceApplicationDocuments",
+    "UserRole": "caseworker-divorce-courtadmin-la",
+    "CRUD": "R"
```
Comment (madjava) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json:735]: This needs to be CRU, would confirm with Jeremy and get the AC's updated

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json [hunk 12]
```diff
@@ -705,5 +705,47 @@
     "CaseFieldID": "LabelContinueBecauseOfDispensedAccepted",
     "UserRole": "caseworker-divorce-solicitor",
     "CRUD": "R"
+  },
+  {
+    "LiveFrom": "03/09/2020",
```
Comment (yarpo) [inline definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json:710]: is this so old? :)

File: definitions/divorce/json/ComplexTypes/ComplexTypes-deemed-and-dispensed-nonprod.json (hunk 6 above)
Comment (yarpo) [reply definitions/divorce/json/ComplexTypes/ComplexTypes-deemed-and-dispensed-nonprod.json:10]: if you need to store just `DivorceDocument` there is no need of defining your own complex type.

You can use:

```
{
    "LiveFrom": "24/09/2020",
    "CaseTypeID": "DIVORCE",
    "ID": "ServiceApplicationDocuments",
    "Label": "Service Application Documents",
    "FieldType": "Collection",
    "FieldTypeParameter": "DivorceDocument",
    "SecurityClassification": "Public"
  },
```

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json (hunk 12 above)
Comment (madjava) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json:710]: @yarpo Would go though LiveFrom date's and update, somehow it always skips me :)

File: test/unit/definitions/divorce/CaseTypeTab.test.js [hunk 13]
```diff
@@ -33,117 +48,145 @@ describe('CaseTypeTab', () => {
...
+    const validationErrors = validateUniqueTabDisplayOrder(tabIds, caseTypeTab);
+    expect(validationErrors).to.have.lengthOf(0);
+  });
+
+  it('should contain proper sequence for TabFieldDisplayOrder with no gaps', () => {
+    assert.doesNotThrow(() => {
+      validateTabFieldDisplayOrder(tabIds, caseTypeTab);
+    },
+    /Missing\/unordered TabFieldDisplayOrder sequence number in TabID/);
+  });
```
Comment (madjava) [inline test/unit/definitions/divorce/CaseTypeTab.test.js:191]: Reusing common functionality

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json (hunk 1 above)
Comment (Dizrupto) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json:749]: I think you can add "CRU" to caseworker-beta also.
Also, not sure why, but it seems the existing document collections event put "CRUD" for all the roles. Not sure whether there is a specific reason for that.

**Review by Dizrupto — APPROVED** [review]

File: definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json (hunk 2 above)
Comment (yarpo) [reply definitions/divorce/json/CaseField/CaseField-deemed-and-dispensed-nonprod.json:10]: I think this should be just a collection of complex type <DivorceDocument>. 

Pretty much the same as:

```
  {
    "LiveFrom": "01/01/2017",
    "CaseTypeID": "DIVORCE",
    "ID": "D8DocumentsGenerated",
    "Label": "Documents generated",
    "FieldType": "Collection",
    "FieldTypeParameter": "DivorceDocument",
    "SecurityClassification": "Public"
  },
```

File: definitions/divorce/json/ComplexTypes/ComplexTypes-deemed-and-dispensed-nonprod.json (hunk 6 above)
Comment (madjava) [reply definitions/divorce/json/ComplexTypes/ComplexTypes-deemed-and-dispensed-nonprod.json]: @yarpo Updated

File: definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json (hunk 1 above)
Comment (madjava) [reply definitions/divorce/json/AuthorisationCaseField/AuthorisationCaseField-deemed-and-dispensed-nonprod.json]: @Dizrupto Thanks, noticed same so i have added `CRUD` for these roles and included `caseworker-beta` also

**Review by yarpo — APPROVED** [review]

**Pull request merged**